### ISSUE TITLE
Send failures to peer should be recorded, and nodes eventually removed from peer list if unreachable

### ISSUE THREAD

**Issue by alfwatt** [tendermint/tendermint#154]
On starting a node I find that one of the peers is not accessible and see repeated log messages like:

```
WARN[10-14|07:14:39] Send failed                              module=p2p channel=20 conn=MConn{125.16.134.94:49160} msg="[HasVote VI:2 V:{42397/00/2} VI:2]"
```

Some failures may be transient or expected, but a peer which is consistently un-reachable should be removed from the peer list.

Potential strategies included keeping a success/failure counters for each peer and making desiccation based on the ratio, but this would potentially lock peers out after a temporary network disconnection.

Suggest keeping an 'outstanding failure' counter which increments on each failure and decrements on each success, using '0' as the baseline for a perfectly reliable peer and a configurable cap on the number or failures that will be tolerated before dropping the peer (~100 seems like a reasonable first value).

Rate limiting of messages to peers with an outstanding failure is also an option, by adding a 'last contact' date field per peer, and using the time difference to determine if a node should attempt to contact the peer.

**Comment by jaekwon** [comment]
Nice, thanks Alf!

Did you happen to see any messages being sent to `125.16.134.94:49160`? There are ping/pong messages in the p2p/connection.go implementation which should have caught dead connections.  My guess is that the connection to 125.16.134.94 was really slow and the send queue was filling up occasionally, but not all the time.  Does that seem right?

Your suggestions seem good, will implement, but also curious if it was a slow connection or a dead one.

**Comment by alfwatt** [comment]
I only saw connection failures and a lot of them. The terminal app I was using at the time truncated the log so I can’t search back for any earlier ping messages, I’ll start it up again with a log file and post that…

Worth considering the various proposed solutions with respect to potentially fragmenting the network because one node can’t see another, it’s probably not a good idea to allow one node to report on another’s unreachability so that a misbehaved or poorly connected node doesn’t cause the rest of the network to start ignoring a particular peer.

**Comment by adrianbrink** [comment]
This got resolved with exponential retries in abc3a2c
